src/Nethermind/Nethermind.Optimism/OptimismPayloadPreparationService.cs

```diff
@@ -45,6 +46,29 @@ public OptimismPayloadPreparationService(
     protected override void ImproveBlock(string payloadId, BlockHeader parentHeader,
         PayloadAttributes payloadAttributes, Block currentBestBlock, DateTimeOffset startDateTime)
     {
+        if (payloadAttributes is OptimismPayloadAttributes optimismPayload)
+        {
+            var spec = _specProvider.GetSpec(currentBestBlock.Header);
+            if (spec.IsOpHoloceneEnabled)
+            {
+                if (!optimismPayload.TryDecodeEIP1559Parameters(out EIP1559Parameters eip1559Parameters, out var error))
+                {
+                    throw new InvalidOperationException($"{nameof(OptimismPayloadAttributes)} was not properly validated: invalid {nameof(OptimismPayloadAttributes.EIP1559Params)}");
+                }
+
+                if (eip1559Parameters.IsZero())
+                {
+                    eip1559Parameters = new EIP1559Parameters(eip1559Parameters.Version, (UInt32)spec.BaseFeeMaxChangeDenominator, (UInt32)spec.ElasticityMultiplier);
+                }
+
+                currentBestBlock.Header.ExtraData = new byte[EIP1559Parameters.ByteLength];
+                eip1559Parameters.WriteTo(currentBestBlock.Header.ExtraData);
+
+                // NOTE: Since we updated the `Header` we need to recalculate the hash.
+                currentBestBlock.Header.Hash = currentBestBlock.Header.CalculateHash();
```

Comment:
As the comment highlights, since we're modifying the `BlockHeader.ExtraData` we need to recompute the `Hash`. This is again error prone and tedious since we can easily forget to update the hash.

It's not clear to me if this is the best place to perform this recalculation or maybe there is a different stage in the processing pipeline where it should be done.